include/seastar/core/semaphore.hh

```diff
@@ -323,18 +357,41 @@ basic_semaphore<ExceptionFactory, Clock>::broken(std::exception_ptr xp) {
     }
 }
 
+#ifdef SEASTAR_DEADLOCK_DETECTION
+template<typename ExceptionFactory, typename Clock>
+#else
 template<typename ExceptionFactory = semaphore_default_exception_factory, typename Clock = typename timer<>::clock>
+#endif
 class semaphore_units {
     basic_semaphore<ExceptionFactory, Clock>* _sem;
     size_t _n;
-
-    semaphore_units(basic_semaphore<ExceptionFactory, Clock>* sem, size_t n) noexcept : _sem(sem), _n(n) {}
+#ifdef SEASTAR_DEADLOCK_DETECTION
+    task* _prev_tsk;
+    task* _cur_tsk;
+#endif
+    semaphore_units(basic_semaphore<ExceptionFactory, Clock>* sem, size_t n) noexcept : _sem(sem), _n(n) {
+    #ifdef SEASTAR_DEADLOCK_DETECTION
+        _cur_tsk = current_task();
+        _sem->add_units(this);
+    #endif
+    }
+#ifdef SEASTAR_DEADLOCK_DETECTION
+    void update_cur_tsk() {
+        if (!_cur_tsk) {
```

Comment:
We might want to introduce here some consistency check, like to verify that if `_cur_tsk` is either null, or equal to `current_task()`.